ros2bag/ros2bag/verb/record.py

```diff
@@ -60,6 +60,11 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
             '-x', '--exclude', default='',
             help='Exclude topics containing provided regular expression. '
             'Works on top of --all, --regex, or topics list.')
+        parser.add_argument(
+            '--include-unpublished-topics', action='store_true',
```

Comment:
@theseankelly nit^2: would `--wait-for-publishers` be a more accurate name for this flag?

Reply:
The new behavior (don't add unpublished topics to the bag) is *off* by default in the current implementation. Naming it `--wait-for-publishers` would imply that the new behavior should be *on* by default.

Given the description in the underlying issue report, I feel like the most commonly desired behavior is to *not* record bags until they are published so that we avoid any QoS incompatibilities.

I could be persuaded if there's strong desire -- I do realize this change alters the default behavior of `ros2 bag record -a` w.r.t unpublished topics.

Reply:
> The new behavior (don't add unpublished topics to the bag) is off by default in the current implementation. Naming it --wait-for-publishers would imply that the new behavior should be on by default.

Yeah, fair point.

> I could be persuaded if there's strong desire -- I do realize this change alters the default behavior of `ros2 bag record -a` w.r.t unpublished topics.

That's true, but I agree that the current default behavior is less than intuitive at best and harmful at worst (e.g. if `/tf_static` messages are silently dropped). 

It does beg the question though, what would be the use case for early subscription with default QoS policies? Grabbing the first few messages sent by a publisher? If there's no other subscriber from which to derive the topic type, DDS discovery latency and graph polling will make it impossible. CC @emersonknapp for feedback.

Reply:
I'm also interested in hearing what @emersonknapp has to say.

I think there are a couple (maybe contrived..) scenarios where the feature has value. For example, spinning up a background rosbag snapshotter as part of a robot topology - it might be useful to not have to worry about boot order of the various nodes in the system.

Definitely niche, and might only be useful with a couple other flags (`--qos-profiles-override-path` and `--no-discovery` or something).

Reply:
@emersonknapp Gentle ping in this issue if you have any concerns or suggestions to the proposed solution it's good time to express it. 
Otherwise if CI will be green I think we will merge it Monday or Tuesday.

IMO alternative scenario described in #967 is a valid case which should be handled correctly
```
1. In terminal 1 create a subscriber by running ros2 topic echo --qos-reliability best_effort /foo std_msgs/msg/Int32
2. In terminal 2 run ros2 record -a
    a. NOTE in the output that the topic /foo was added to the bag -- found by the DataReader
3. In terminal 3 create a publisher via ros2 topic pub --qos-reliability best_effort /foo std_msgs/msg/Int32 '{value: 1}'

The DataWriter (and its offered QoS) will be discovered, and terminal 2 (rosbag record) will display the warning about incompatible publishers.
```

Reply:
Sorry for late reply - my perspective is that "will not create a subscription until at least one publisher is discovered" should be the default behavior of discovery  That seems more intelligent behavior overall. In fact, I'm not sure I see any reason to turn it off, unless we're worried that the discovery loop is a performance overhead that a user may want to explicitly disable?

Even in that case, the default QoS is kind of an arbitrary baseline subscription to fall back to. It would probably make more sense that "if you provide a QoS override for an explicitly specified subscription topic, then the subscription will be created immediately using that override and will not impact discovery".

Given that, I'm not sure we need a new commandline option for this feature! I'll note that this is a good thing probably, we already have an unwieldy number of options.

Reply:
@emersonknapp  I also don't envision the case when we might need to use `--include-unpublished-topics` parameter.   

@theseankelly  Do you have any scenario in mind or in practice when we might need to use `--include-unpublished-topics`?  
If not I would ask you to remove this extra parameter from command line option.

IMO Possible concern that the discovery thread affecting performance doesn't count. Since it's very minor and used to be only for the begging of the recording.

I agree with @emersonknapp that 
> "if you provide a QoS override for an explicitly specified subscription topic, then the subscription will be created immediately using that override and will not use discovery".

However I would moved it as follow up feature request for the sake of saving time to merge current PR which is considering as a "bugfix".